File: architecture-skills/design-patterns/examples/adapter/adapter.swift
// Adapter pattern in Swift.
//
// Swift's idiomatic adapter is a retroactive protocol conformance: an
// extension makes an existing (even third-party) type conform to the
// target protocol with no wrapper type at all. Fall back to a wrapper
// struct when the conformance needs extra state or would clash with
// the type's own semantics.

import Foundation

protocol TemperatureSource {
    var name: String { get }
    var celsius: Double { get }
}

struct ModernSensor: TemperatureSource {
    let name: String
    let celsius: Double
}

// Adaptee: imagine this class comes from a third-party module.
final class LegacyFahrenheitSensor {
    let sensorId: String
    private let fahrenheit: Double

    init(sensorId: String, fahrenheit: Double) {
        self.sensorId = sensorId
        self.fahrenheit = fahrenheit
    }

    func readFahrenheit() -> Double { fahrenheit }
}

// The extension *is* the adapter.
extension LegacyFahrenheitSensor: TemperatureSource {
    var name: String { sensorId }
    var celsius: Double { (readFahrenheit() - 32.0) * 5.0 / 9.0 }
}

func report(_ sources: [TemperatureSource]) {
    for source in sources {
        print("\(source.name): \(String(format: "%.1f", source.celsius))°C")
    }
}

report([
    ModernSensor(name: "server-room", celsius: 21.5),
    LegacyFahrenheitSensor(sensorId: "warehouse-legacy", fahrenheit: 77.0),
])

File: architecture-skills/design-patterns/examples/adapter/adapter.rs
// Adapter pattern in Rust.
//
// The newtype idiom is the natural adapter: wrap the foreign type in a
// local struct and implement the local trait on the wrapper. This also
// sidesteps the orphan rule, which forbids implementing a foreign trait
// for a foreign type directly.

trait TemperatureSource {
    fn name(&self) -> &str;
    fn celsius(&self) -> f64;
}

struct ModernSensor {
    name: String,
    celsius: f64,
}

impl TemperatureSource for ModernSensor {
    fn name(&self) -> &str {
        &self.name
    }
    fn celsius(&self) -> f64 {
        self.celsius
    }
}

/// Adaptee: imagine this type comes from another crate.
struct LegacyFahrenheitSensor {
    sensor_id: String,
    fahrenheit: f64,
}

impl LegacyFahrenheitSensor {
    fn read_fahrenheit(&self) -> f64 {
        self.fahrenheit
    }
}

/// Newtype adapter around the legacy sensor.
struct FahrenheitAdapter(LegacyFahrenheitSensor);

impl TemperatureSource for FahrenheitAdapter {
    fn name(&self) -> &str {
        &self.0.sensor_id
    }
    fn celsius(&self) -> f64 {
        (self.0.read_fahrenheit() - 32.0) * 5.0 / 9.0
    }
}

fn report(sources: &[&dyn TemperatureSource]) {
    for source in sources {
        println!("{}: {:.1}°C", source.name(), source.celsius());
    }
}

fn main() {
    let modern = ModernSensor {
        name: "server-room".to_string(),
        celsius: 21.5,
    };
    let adapter = FahrenheitAdapter(LegacyFahrenheitSensor {
        sensor_id: "warehouse-legacy".to_string(),
        fahrenheit: 77.0,
    });
    report(&[&modern, &adapter]);
}
